### ISSUE TITLE
Make sure Azle works well with npm workspaces

### ISSUE THREAD

**Issue by lastmjs** [demergent-labs/azle#2285]
Right now dfx deploy of an azle canister only works if there is a node_modules directory in the same directory as the directory from which you execute `dfx deploy`. We want to support a variety of different directory structures, and most notably `npm workspaces`. Please do that. You might consider using `npx`, `npm exec --offline` etc. Make sure that this command will not try to go to the internet and globally install azle, that's why we switched to the `node_modules/azle` command in our dfx extension in the first place.

- [x] Make sure we understand the limitations of workspaces
     - [x] if the dfx.json and package.json aren't siblings will it still work?
          - [x] Make sure to check if the package.json is in an ancestor dir and descendant dir of the dfx.json dir
- [x] have a version of azle in the custom 
     - [x] `build": "npx exec azle@{{custom.azle_version}} compile {{canister_name}}",`
     - [x] We don't need this anymore
- [x] Test everything that we test with npm link outside of our repo as well, like on a release candidate
- [x] see if we can hide the true command that jordan thinks is ugly from the dfx build
     - [x] compare the above command to how it looks in rust and motoko
          - `Executing: cargo build --target wasm32-unknown-unknown --release -p my_rust_example_backend --locked`
          - Nothing for motoko
     - [x] I don't think we can do it, most of the stuff we have in there is so that we can find the right command to run in the first place, and then once we run the command, that it will run as we expect. If we try and hide it somewhere else, then by the time we find it, we will either have had to put all of this information on that same "simplified" command, thus negating the simplification, or we will not have the information we need. Rusts is, imho, just as esoteric. Motoko's is none existent.
- [x] How many commands won't be able to run now without having the dfx variable explicitly set?
     - [x] switch to dfx path constant to a function so it's only run when you need the dfx path and therefore only when you need to have the env variable set
- [x] Add the crazy workspace example into our tests
     - [x] Add env variable that will report the version of azle
     - [x] Find a way to make sure that it's not using a global azle?
          -  using the --prefix flag with npm will make sure that it uses the right azle for that package

**Comment by bdemann** [comment]
okay so here is what seems to be the problem with workspaces. the npm_modules directory is no longer in the test folder, it's in the the workspace folder. However all of the scripts are still run in the test dir, which is no longer the home of the node_modules dir.

**Comment by bdemann** [comment]
tl;dr the solution we came up with was to use the --prefix option for npm which will give it an indication of where it should start looking when resolving packages.
